File: Management/Models/CanvasModels/Courses/CourseModel.cs
using CanvasModel.Enrollments;

namespace CanvasModel.Courses;

public record CourseModel(
  [property: JsonPropertyName("id")] ulong Id,
  [property: JsonPropertyName("sis_course_id")] string SisCourseId,
  [property: JsonPropertyName("uuid")] string Uuid,
  [property: JsonPropertyName("integration_id")] string IntegrationId,
  [property: JsonPropertyName("name")] string Name,
  [property: JsonPropertyName("course_code")] string CourseCode,
  [property: JsonPropertyName("workflow_state")] string WorkflowState,
  [property: JsonPropertyName("account_id")] ulong AccountId,
  [property: JsonPropertyName("root_account_id")] ulong RootAccountId,
  [property: JsonPropertyName("enrollment_term_id")] ulong EnrollmentTermId,
  [property: JsonPropertyName("created_at")] DateTime CreatedAt,
  [property: JsonPropertyName("locale")] string Locale,
  [property: JsonPropertyName("calendar")] CalendarLinkModel Calendar,
  [property: JsonPropertyName("default_view")] string DefaultView,
  [property: JsonPropertyName("syllabus_body")] string SyllabusBody,
  [property: JsonPropertyName("permissions")] Dictionary<string, bool> Permissions,
  [property: JsonPropertyName("storage_quota_mb")] ulong StorageQuotaMb,
  [property: JsonPropertyName("storage_quota_used_mb")] ulong StorageQuotaUsedMb,
  [property: JsonPropertyName("license")] string License,
  [property: JsonPropertyName("course_format")] string CourseFormat,
  [property: JsonPropertyName("time_zone")] string TimeZone,
  [property: JsonPropertyName("sis_import_id")] ulong? SisImportId = null,
  [property: JsonPropertyName("grading_standard_id")] ulong? GradingStandardId = null,
  [property: JsonPropertyName("start_at")] DateTime? StartAt = null,
  [property: JsonPropertyName("end_at")] DateTime? EndAt = null,
  [property: JsonPropertyName("enrollments")] IEnumerable<EnrollmentModel>? Enrollments = null,
  [property: JsonPropertyName("total_students")] ulong? TotalStudents = null,
  [property: JsonPropertyName("needs_grading_count")] uint? NeedsGradingCount = null,
  [property: JsonPropertyName("term")] TermModel? Term = null,
  [property: JsonPropertyName("course_progress")] CourseProgressModel? CourseProgress = null,
  [property: JsonPropertyName("apply_assignment_group_weights")]
    bool? ApplyAssignmentGroupWeights = null,
  [property: JsonPropertyName("is_public")] bool? Is = null,
  [property: JsonPropertyName("is_public_to_auth_users")] bool? IsPublicToAuthUsers = null,
  [property: JsonPropertyName("public_syllabus")] bool? PublicSyllabus = null,
  [property: JsonPropertyName("public_syllabus_to_auth")] bool? PublicSyllabusToAuth = null,
  [property: JsonPropertyName("public_description")] string? PublicDescription = null,
  [property: JsonPropertyName("hide_final_grades")] bool? HideFinalGrades = null,
  [property: JsonPropertyName("allow_student_assignment_edits")]
    bool? AllowStudentAssignmentEdits = null,
  [property: JsonPropertyName("allow_wiki_comments")] bool? AllowWikiComments = null,
  [property: JsonPropertyName("allow_student_forum_attachments")]
    bool? AllowStudentForumAttachments = null,
  [property: JsonPropertyName("open_enrollment")] bool? OpenEnrollment = null,
  [property: JsonPropertyName("self_enrollment")] bool? SelfEnrollment = null,
  [property: JsonPropertyName("restrict_enrollments_to_courses")]
    bool? RestrictEnrollmentsToCourseDates = null,
  [property: JsonPropertyName("access_restricted_by_date")] bool? AccessRestrictedByDate = null,
  [property: JsonPropertyName("blueprint")] bool? Blueprint = null,
  [property: JsonPropertyName("blueprint_restrictions")]
    Dictionary<string, bool>? BlueprintRestrictions = null,
  [property: JsonPropertyName("blueprint_restrictions_by_object_type")]
    Dictionary<string, Dictionary<string, bool>>? BlueprintRestrictionsByObjectType = null
);

File: nextjs/src/models/canvas/courses/courseModel.ts
import { EnrollmentModel } from "../enrollments/enrollmentModel";
import { CalendarLinkModel } from "./calendarLinkModel";
import { CourseProgressModel } from "./courseProgressModel";
import { TermModel } from "./termModel";

export interface CourseModel {
  id: number;
  sis_course_id: string;
  uuid: string;
  integration_id: string;
  name: string;
  course_code: string;
  workflow_state: string;
  account_id: number;
  root_account_id: number;
  enrollment_term_id: number;
  created_at: string; // ISO 8601 date string
  locale: string;
  calendar: CalendarLinkModel;
  default_view: string;
  syllabus_body: string;
  permissions: { [key: string]: boolean };
  storage_quota_mb: number;
  storage_quota_used_mb: number;
  license: string;
  course_format: string;
  time_zone: string;
  sis_import_id?: number;
  grading_standard_id?: number;
  start_at?: string; // ISO 8601 date string
  end_at?: string; // ISO 8601 date string
  enrollments?: EnrollmentModel[];
  total_students?: number;
  needs_grading_count?: number;
  term?: TermModel;
  course_progress?: CourseProgressModel;
  apply_assignment_group_weights?: boolean;
  is_public?: boolean;
  is_public_to_auth_users?: boolean;
  public_syllabus?: boolean;
  public_syllabus_to_auth?: boolean;
  public_description?: string;
  hide_final_grades?: boolean;
  allow_student_assignment_edits?: boolean;
  allow_wiki_comments?: boolean;
  allow_student_forum_attachments?: boolean;
  open_enrollment?: boolean;
  self_enrollment?: boolean;
  restrict_enrollments_to_course_dates?: boolean;
  access_restricted_by_date?: boolean;
  blueprint?: boolean;
  blueprint_restrictions?: { [key: string]: boolean };
  blueprint_restrictions_by_object_type?: { [key: string]: { [key: string]: boolean } };
}
